[Flow] Add kafka Component (#4110)
grafana/agent: docs/sources/flow/reference/components/prometheus.exporter.kafka.md

---
# NOTE(rfratto): the title below has zero-width spaces injected into it to
# prevent it from overflowing the sidebar on the rendered site. Be careful when
# modifying this section to retain the spaces.
#
# Ideally, in the future, we can fix the overflow issue with css rather than
# injecting special characters.

title: prometheus.exporter.kafka
---

# prometheus.exporter.kafka
The `prometheus.exporter.kafka` component embeds

> [kafka_exporter](https://github.com/davidmparrott/kafka_exporter/v2/exporter) for collecting metrics from a kafka server.

Reviewer: this link gives a 404

Author: Yes, I will modify to the correct address

## Usage

```river
prometheus.exporter.kafka "LABEL" {
    kafka_uris = ["localhost:9200"]
}
```

## Arguments
The following arguments can be used to configure the exporter's behavior.
Omitted fields take their default values.

| Name                        | Type            | Description                                                                                                                             | Default | Required |
|-----------------------------|-----------------|-----------------------------------------------------------------------------------------------------------------------------------------|---------|----------|

> | `instance`                  | `string`        |                                                                                                                                         |         | no       |
> | `kafka_uris`                | `array(string)` | Address array (host:port) of Kafka server.                                                                                              |         | yes      |

Reviewer 2: nit: I think that the required arguments should be at the top of the list.
```suggestion
| `kafka_uris`                | `array(string)` | Address array (host:port) of Kafka server.                                                                                              |         | yes      |
| `instance`                  | `string`        |                                                                                                                                         |         | no       |
```

| `targets` | `list(map(string))` | The targets that can be used to collect `kafka` metrics. |

For example, `targets` can either be passed to a `prometheus.relabel`
component to rewrite the metrics' label set, or to a `prometheus.scrape`
component that collects the exposed metrics.

The exported targets will use the configured [in-memory traffic][] address
specified by the [run command][].

[in-memory traffic]: {{< relref "../../concepts/component_controller.md#in-memory-traffic" >}}
[run command]: {{< relref "../cli/run.md" >}}

## Component health

`prometheus.exporter.kafka` is only reported as unhealthy if given
an invalid configuration. In those cases, exported fields retain their last
healthy values.

## Debug information

`prometheus.exporter.kafka` does not expose any component-specific
debug information.

## Debug metrics

`prometheus.exporter.kafka` does not expose any component-specific
debug metrics.

## Example

This example uses a [`prometheus.scrape` component][scrape] to collect metrics
from `prometheus.exporter.kafka`:

```river
prometheus.exporter.kafka "example" {
  kafka_uris = ["localhost:9200"]
}

> // Configure a prometheus.scrape component to collect kafka metrics.
> prometheus.scrape "demo" {
>   targets    = prometheus.exporter.kafka.example.targets
>   forward_to = [ /* ... */ ]
> }

Reviewer 2: Ideally examples are as copy-paste-able as we can make them:
```suggestion
// Configure a prometheus.scrape component to send metrics to
// prometheus.remote_write component.
prometheus.scrape "demo" {
  targets = [
    prometheus.exporter.kafka.example.targets
  ]
  forward_to = [prometheus.remote_write.demo.receiver]
}

prometheus.remote_write "demo" {
  endpoint {
    url = "http://mimir:9009/api/v1/push"

    basic_auth {
      username = "example-user"
      password = "example-password"
    }
  }
}
```

---
# NOTE(rfratto): the title below has zero-width spaces injected into it to
# prevent it from overflowing the sidebar on the rendered site. Be careful when
# modifying this section to retain the spaces.
#
# Ideally, in the future, we can fix the overflow issue with css rather than
# injecting special characters.

title: prometheus.exporter.kafka
---

# prometheus.exporter.kafka
The `prometheus.exporter.kafka` component embeds
[kafka_exporter](https://github.com/davidmparrott/kafka_exporter) for collecting metrics from a kafka server.

## Usage

```river
prometheus.exporter.kafka "LABEL" {
    kafka_uris = ["localhost:9200"]
}
```

## Arguments

> The following arguments can be used to configure the exporter's behavior.

Reviewer 3: ```suggestion
You can use the following arguments to configure the exporter's behavior.
```

Make it clear who can use the arguments and make sentence active.

Omitted fields take their default values.

| Name                        | Type            | Description                                                                                                                                                                         | Default | Required |
|-----------------------------|-----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|----------|
| `kafka_uris`                | `array(string)` | Address array (host:port) of Kafka server.                                                                                                                                          |         | yes      |

> | `instance`                  | `string`        | The`instance`label for metrics, default is the hostname:port of the first kafka_uris, If there is more than one string in kafka_uris, the instance value must be manually provided. |         | no       |

Reviewer 3: ```suggestion
| `instance`                  | `string`        | The`instance`label for metrics, default is the hostname:port of the first kafka_uris. If there is more than one string in kafka_uris, you must manually provide the instance value. |         | no       |
```

| `use_tls`                   | `bool`          | Connect using TLS.                                                                                                                                                                 |         | no       |
| `ca_file`                   | `string`        | The optional certificate authority file for TLS client authentication.                                                                                                             |         | no       |
| `cert_file`                 | `string`        | The optional certificate file for TLS client authentication.                                                                                                                       |         | no       |
| `key_file`                  | `string`        | The optional key file for TLS client authentication.                                                                                                                               |         | no       |

> | `insecure_skip_verify`      | `bool`          | If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure.                                                                |         | no       |

Reviewer 3: ```suggestion
| `insecure_skip_verify`      | `bool`          | If set to true, the server's certificate will not be checked for validity. This makes your HTTPS connections insecure.                                                                |         | no       |
```

| `kafka_version`             | `string`        | Kafka broker version.                                                                                                                                                              | `2.0.0` | no       |

> | `use_zookeeper_lag`         | `bool`          | if you need to use a group from zookeeper.                                                                                                                                         |         | no       |

Reviewer 3: ```suggestion
| `use_zookeeper_lag`         | `bool`          | If set to true, use a group from zookeeper.                                                                                                                                         |         | no       |
```

> | `zookeeper_uris`            | `array(string)` | Address array (hosts) of zookeeper server..                                                                                                                                        |         | no       |

Reviewer 3: ```suggestion
| `zookeeper_uris`            | `array(string)` | Address array (hosts) of zookeeper server.                                                                                                                                        |         | no       |
```

| `kafka_cluster_name`        | `string`        | Kafka cluster name.                                                                                                                                                                 |         | no       |
| `metadata_refresh_interval` | `duration`      | Metadata refresh interval.                                                                                                                                                          | `1m`    | no       |

> | `allow_concurrency`         | `bool`          | If true, all scrapes will trigger kafka operations otherwise, they will share results. WARN: This should be disabled on large clusters.                                             | `true`  | no       |

Reviewer 3: ```suggestion
| `allow_concurrency`         | `bool`          | If set to true, all scrapes trigger Kafka operations. Otherwise, they will share results. WARNING: Disable this on large clusters.                                            | `true`  | no       |
```

> | `max_offsets`               | `int`           | Maximum number of offsets to store in the interpolation table for a partition.                                                                                                      | `1000`  | no       |

Reviewer 3: ```suggestion
| `max_offsets`               | `int`           | The maximum number of offsets to store in the interpolation table for a partition.                                                                                                      | `1000`  | no       |
```

| `prune_interval_seconds`    | `int`           | How frequently should the interpolation table be pruned, in seconds.                                                                                                                | `30`    | no       |
| `topics_filter_regex`       | `string`        | Regex filter for topics to be monitored.                                                                                                                                            | `.*`    | no       |
| `groups_filter_regex`       | `string`        | Regex filter for consumer groups to be monitored.                                                                                                                                   | `.*`    | no       |

## Exported fields
The following fields are exported and can be referenced by other components.

| Name      | Type                | Description                                              |
|-----------|---------------------|----------------------------------------------------------|

> | `targets` | `list(map(string))` | The targets that can be used to collect `kafka` metrics. |

Reviewer 3: Why is Kafka in code here? And not just Kafka?

Author: Yes, It's only collect kafka metrics, So the word `kafka` is unnecessary here

Reviewer 3: So should the text here be: 

> The targets that can be used to collect Kafka metrics.

(no code syntax, and capitalise Kafka)
